edgelet/iotedged/src/lib.rs

```diff
@@ -759,6 +1003,104 @@ fn manual_provision(
+    hyper_client: HC,
+    backup_path: PathBuf,
+    runtime: M,
+    tokio_runtime: &mut tokio::runtime::Runtime,
+    hybrid_identity_key: &[u8],
+    cert_thumbprint: String,
+) -> Result<(DerivedKeyStore<MemoryKey>, ProvisioningResult, MemoryKey, M), Error>
+where
+    HC: 'static + ClientImpl,
+    M: ModuleRuntime + Send + 'static,
+{
+    let mut memory_hsm = MemoryKeyStore::new();
+
+    memory_hsm
+        .activate_identity_key(
+            KeyIdentity::Device,
+            "primary".to_string(),
+            hybrid_identity_key,
+        )
+        .context(ErrorKind::ActivateSymmetricKey)?;
+
+    let dps = DpsX509Provisioning::new(
+        hyper_client,
+        provisioning.global_endpoint().clone(),
+        provisioning.scope_id().to_string(),
+        reg_id,
+        DPS_API_VERSION.to_string(),
+    )
+    .context(ErrorKind::Initialize(
+        InitializeErrorReason::DpsProvisioningClient,
+    ))?;
+
+    let provision_with_file_backup = BackupProvisioning::new(dps, backup_path);
+
+    let provision = provision_with_file_backup
+        .provision(memory_hsm.clone())
+        .map_err(|err| {
+            Error::from(err.context(ErrorKind::Initialize(
+                InitializeErrorReason::DpsProvisioningClient,
+            )))
+        })
+        .and_then(move |prov_result| {
+            info!("Successful DPS provisioning.");
+            if prov_result.reconfigure() == ReprovisioningStatus::DeviceDataNotUpdated {
+                Either::B(future::ok((prov_result, runtime)))
+            } else {
+                info!(
+                    "Reprovisioning status {:?} will trigger reconfiguration of modules.",
+                    prov_result.reconfigure()
+                );
+
+                let remove = runtime.remove_all().then(|result| {
+                    result.context(ErrorKind::Initialize(
+                        InitializeErrorReason::DpsProvisioningClient,
+                    ))?;
+                    Ok((prov_result, runtime))
+                });
+                Either::A(remove)
+            }
+        })
+        .and_then(move |(prov_result, runtime)| {
+            let (derived_key_store, hybrid_derived_key) = prepare_hybrid_key(
+                &memory_hsm,
+                &cert_thumbprint,
+                prov_result.hub_name(),
+                prov_result.device_id(),
+            )?;
+            Ok((derived_key_store, prov_result, hybrid_derived_key, runtime))
+        });
+    tokio_runtime.block_on(provision)
+}
+
+fn prepare_hybrid_key(
+    key_store: &MemoryKeyStore,
+    cert_thumbprint: &str,
+    hub_name: &str,
+    device_id: &str,
+) -> Result<(DerivedKeyStore<MemoryKey>, MemoryKey), Error> {
+    let k = key_store
+        .get(&KeyIdentity::Device, "primary")
+        .context(ErrorKind::Initialize(
+            InitializeErrorReason::DpsProvisioningClient,
+        ))?;
+    let sign_data = format!("{}{}{}", hub_name, device_id, cert_thumbprint);
```

Comment:
It doesn't really matter, but maybe separate the components with a `.` like in JWK ?

Reply:
Sure